File: trunk/MS.NET/clrzbus/zbus/zmsg.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;
using System.Runtime.InteropServices;
 
namespace zbus {  
 
    public class ZMsg{
        private LinkedList<byte[]> frames = new LinkedList<byte[]>();
        private long contentSize = 0;

        public ZMsg()
        { 
        }
 
        public void PushBack(byte[] frame){
            frames.AddLast(frame);
            this.contentSize += frame.Length;
        }

        public void PushBack(string frame, Encoding encoding)
        {
            PushBack(encoding.GetBytes(frame));
        }

        public void PushBack(string frame)
        {
            PushBack(frame, Encoding.UTF8);
        }

        public void PushFront(byte[] frame)
        {
            frames.AddFirst(frame);
            this.contentSize += frame.Length;
        }

        public void PushFront(string frame, Encoding encoding)
        {
            PushFront(encoding.GetBytes(frame));
        }

        public void PushFront(string frame)
        {
            PushFront(frame, Encoding.UTF8);
        } 

        public byte[] PopFront(){
            if(frames.First == null){
                return null;
            }
            byte[] frame = frames.First.Value;
            frames.RemoveFirst();
            this.contentSize -= frame.Length;
            return frame;
        }
        public string PopFrontStr(Encoding encoding)
        {
            byte[] frame = PopFront();
            if (frame == null) return null;
            return encoding.GetString(frame); 
        }

        public string PopFrontStr()
        {
            return PopFrontStr(Encoding.UTF8);
        }


        public byte[] PopBack()
        {
            if (frames.Last == null)
            {
                return null;
            }
            byte[] frame = frames.Last.Value;
            frames.RemoveLast();
            this.contentSize -= frame.Length;
            return frame;
        }
        public string PopBackStr(Encoding encoding)
        {
            byte[] frame = PopBack();
            if (frame == null) return null;
            return encoding.GetString(frame); 
        }

        public string PopBackStr()
        { 
            return PopFrontStr(Encoding.UTF8);
        }


        public IEnumerator<byte[]> GetEnumerator()
        {
            return frames.GetEnumerator();
        }

        public int FrameSize {
            get { return this.frames.Count; }
        }

        public long ContentSize{
            get { return this.contentSize; }
        }

        public void Clear(){
            frames.Clear();
            this.contentSize = 0;
        }
        public byte[] First{
            get{
                if (this.frames.First == null) return null;

                return this.frames.First.Value;
            }
        }

        public byte[] Last{
            get{
                if (this.frames.Last == null) return null;

                return this.frames.Last.Value;
            }
        }


        private String strhex(byte[] data, int offset, int len) {		
		    String hexChar = "0123456789ABCDEF";
		    StringBuilder b = new StringBuilder();
		    for (int nbr = offset; nbr<len; nbr++) {
			    int b1 = (data[nbr]>>4) & 0xf;
			    int b2 = data[nbr] & 0xf;
			    b.Append(hexChar[b1]);
                b.Append(hexChar[b2]);
		    }
            return b.ToString();
	    }

        private String frameString(byte[] data)
        {
            bool isText = true;
            int maxSize = data.Length;
            string elipsis = "";
            if (maxSize > 128) {
                maxSize = 128;
                elipsis = "...";
            }
            for (int i = 0; i < maxSize; i++)
            {
                if (data[i] < 32 || data[i] > 127)
                    isText = false;
            }
            if (isText)
                return Encoding.UTF8.GetString(data, 0, maxSize) + elipsis;
            else
                return strhex(data, 0, maxSize)+elipsis;
        }

        public void Dump(){
		    Console.WriteLine("---------------------------------");
		    foreach(byte[] data in this.frames){
			    Console.WriteLine("[{0,3:D3}] {1}", data.Length, frameString(data));
		    }
	    }

        public static byte[] ToZFrame(IntPtr frame)
        {
            int frameSize = C.zmq_msg_size(frame);
            byte[] data = new byte[frameSize];
            Marshal.Copy(C.zmq_msg_data(frame), data, 0, frameSize);

            C.zframe_destroy(out frame);

            return data;
        }


        public static IntPtr ToZFramePtr(byte[] data)
        {
            IntPtr pdata = C.zmalloc(data.Length);

            Marshal.Copy(data, 0, pdata, data.Length);
            IntPtr frame = C.zframe_new(pdata, data.Length);

            C.zfree(pdata);

            return frame;
        }

        public static ZMsg ToZMsg(IntPtr msg)
        {
            ZMsg zmsg = new ZMsg();
            IntPtr frame = C.zmsg_pop_front(msg);
            while (frame != IntPtr.Zero) {
                int frameSize = C.zmq_msg_size(frame);
                byte[] data = new byte[frameSize];
                Marshal.Copy(C.zmq_msg_data(frame), data, 0, frameSize);
                zmsg.PushBack(data);

                C.zframe_destroy(out frame);
                frame = C.zmsg_pop_front(msg);
            }
            C.zmsg_destroy(out msg);
            return zmsg;
        } 

        public static IntPtr ToZMsgPtr(ZMsg msg){
            IntPtr zmsg = C.zmsg_new();
            IEnumerator<byte[]> fe = msg.GetEnumerator();
            
            while(fe.MoveNext()){
                IntPtr framePtr = ToZFramePtr(fe.Current);
                C.zmsg_push_back(zmsg, framePtr);
            }

            return zmsg;
        }
    }
}

File: trunk/JAVA/jzbus/src/main/java/net/zbus/ZMsg.java
package net.zbus;

import java.util.ArrayDeque;
import java.util.Iterator;
 
public class ZMsg{ 
	private ArrayDeque<byte[]> frames = new ArrayDeque<byte[]>();
	private long contentSize = 0;
	
	public void pushBack(byte[] frame){ 
		this.frames.add(frame);
		this.contentSize += frame.length;
	}
	
	public void pushBack(String frame){
		this.pushBack(frame.getBytes());
	}
	
	public void pushFront(byte[] frame){
		this.frames.push(frame);
		this.contentSize += frame.length;
	}
	
	public void pushFront(String frame){
		this.pushFront(frame.getBytes());
	}
	
	public byte[] popFront(){
		byte[] data = this.frames.pollFirst();
		if(data == null) return null;
		
		this.contentSize -= data.length;
		return data;
	}
	
	public byte[] popBack(){
		byte[] data = this.frames.pollLast();
		if(data == null) return null;
		
		this.contentSize -= data.length;
		return data;
	}
	
	public String popFrontStr(){
		byte[] data = this.popFront();
		if(data == null) return null;
		return new String(data);
	}
	
	public String popBackStr(){
		byte[] data = this.popBack();
		if(data == null) return null;
		return new String(data); 
	}
	
	public int frameSize(){
		return this.frames.size();
	}
	
	public long contentSize(){
		return this.contentSize;
	}
	
	private String strhex(byte[] data, int offset, int len) {		
		String hexChar = "0123456789ABCDEF";
		StringBuilder b = new StringBuilder();
		for (int nbr = offset;nbr<len;nbr++) {
			int b1 = data[nbr] >>> 4 & 0xf;
			int b2 = data[nbr] & 0xf;
			b.append(hexChar.charAt(b1));
			b.append(hexChar.charAt(b2));
		}
		return b.toString();
	}
	
	private String frameString(byte[] data){
		boolean isText = true;
		int maxSize = data.length;
		if(maxSize > 128) maxSize = 128;
		for (int i = 0;i<maxSize;i++) {
			if (data[i] < 32 || data[i] > 127)
				isText = false;
		}
		if (isText) 
			return new String(data, 0, maxSize);
		else
			return strhex(data, 0, maxSize);
	}
	
	public void dump(){
		System.out.println("---------------------------------");
		for(byte[] data : this.frames){
			System.out.printf("[%03d] %s\n", data.length, frameString(data));
		}
	}
	
	public Iterator<byte[]> iterator(){
		return this.frames.iterator();
	}
	
	public void clear(){
		this.frames.clear();
		this.contentSize = 0;
	}
	public byte[] getLast(){
		return this.frames.getLast();
	}
	public byte[] getFirst(){
		return this.frames.getFirst();
	}
}
